### PR TITLE
infoschema: add metrics_summary_by_label table to query all detail metrics

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs <crazycs520@gmail.com>

### What problem does this PR solve? 
Add the `metrics_summary_by_label` table to query all detail metrics.  It is the table `metrics_summary` by label. 
This table can quickly help Users to find abnormal metrics in detail between two different times.

eg:

```sql
>select * from `METRICS_SUMMARY_BY_LABEL` order by sum_value desc;
+--------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------+----------------------------+-----------------------------+-----------------------------+-----------------------------+-----------------------------+
| METRIC_NAME                                      | LABEL                                                                                                                         | TIME                       | SUM_VALUE                   | AVG_VALUE                   | MIN_VALUE                   | MAX_VALUE                   |
+--------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------+----------------------------+-----------------------------+-----------------------------+-----------------------------+-----------------------------+
| tikv_auto_gc_safepoint                           | instance = 127.0.0.1:20180                                                                                                    | 2020-02-06 20:26:51.628000 |           1.58099122657e+13 |           1.58099122657e+12 |           1.58099098657e+12 |           1.58099158656e+12 |
| pd_cluster_status                                | instance, type = 127.0.0.1:2379, storage_capacity                                                                             | 2020-02-06 20:26:51.628000 |           4.99963170816e+12 |           4.99963170816e+11 |           4.99963170816e+11 |           4.99963170816e+11 |
| pd_scheduler_store_status                        | address, instance, store, type = 127.0.0.1:20160, 127.0.0.1:2379, 1, store_capacity                                           | 2020-02-06 20:26:51.628000 |           4.99963170816e+12 |           4.99963170816e+11 |           4.99963170816e+11 |           4.99963170816e+11 |
| tikv_store_size                                  | instance, type = 127.0.0.1:20180, capacity                                                                                    | 2020-02-06 20:26:51.628000 |           4.99963170816e+12 |           4.99963170816e+11 |           4.99963170816e+11 |           4.99963170816e+11 |
| pd_scheduler_store_status                        | address, instance, store, type = 127.0.0.1:20160, 127.0.0.1:2379, 1, store_available                                          | 2020-02-06 20:26:51.628000 |           8.95173033984e+11 | 89517303398.4               | 89205329920.0               | 89868484608.0               |
| tikv_store_size                                  | instance, type = 127.0.0.1:20180, available                                                                                   | 2020-02-06 20:26:51.628000 |           8.94733672448e+11 | 89473367244.8               | 89183318016.0               | 89855328256.0               |
| pd_scheduler_store_status                        | address, instance, store, type = 127.0.0.1:20160, 127.0.0.1:2379, 1, region_score                                             | 2020-02-06 20:26:51.628000 | 10736564536.5               |  1073656453.65              |  1073656118.73              |  1073656751.17              |
| tikv_allocator_stats                             | instance, type = 127.0.0.1:20180, mapped                                                                                      | 2020-02-06 20:26:51.628000 |  2642391040.0               |   264239104.0               |   263536640.0               |   264978432.0               |
| tikv_allocator_stats                             | instance, type = 127.0.0.1:20180, resident                                                                                    | 2020-02-06 20:26:51.628000 |  1942179840.0               |   194217984.0               |   193515520.0               |   194957312.0               |
| tikv_allocator_stats                             | instance, type = 127.0.0.1:20180, active                                                                                      | 2020-02-06 20:26:51.628000 |  1800548352.0               |   180054835.2               |   179871744.0               |   180322304.0               |
| tikv_allocator_stats                             | instance, type = 127.0.0.1:20180, allocated                                                                                   | 2020-02-06 20:26:51.628000 |  1650098608.0               |   165009860.8               |   164888600.0               |   165097960.0               |
| heap_mem_usage                                   | instance, job = 127.0.0.1:10081, tidb                                                                                         | 2020-02-06 20:26:51.628000 |  1494873992.0               |   149487399.2               |   127208008.0               |   168826440.0               |
| heap_mem_usage                                   | instance, job = localhost:9090, prometheus                                                                                    | 2020-02-06 20:26:51.628000 |  1110194840.0               |   111019484.0               |   103603152.0               |   118016312.0               |
| pd_scheduler_store_status                        | address, instance, store, type = 127.0.0.1:20160, 127.0.0.1:2379, 1, store_used                                               | 2020-02-06 20:26:51.628000 |   654853972.0               |    65485397.2               |    65476826.0               |    65493634.0               |
| pd_cluster_status                                | instance, type = 127.0.0.1:2379, storage_size                                                                                 | 2020-02-06 20:26:51.628000 |   654853972.0               |    65485397.2               |    65476826.0               |    65493634.0               |
| tikv_engine_size                                 | instance, type, db = 127.0.0.1:20180, default, raft                                                                           | 2020-02-06 20:26:51.628000 |   576435976.0               |    57643597.6               |    57630460.0               |    57655804.0               |
| tikv_memtable_size                               | cf, instance, type, db = default, 127.0.0.1:20180, mem-tables, raft                                                           | 2020-02-06 20:26:51.628000 |   571758336.0               |    57175833.6               |    57162696.0               |    57188040.0               |
| tikv_engine_size                                 | instance, type, db = 127.0.0.1:20180, write, kv                                                                               | 2020-02-06 20:26:51.628000 |   346880962.0               |    34688096.2               |    34686165.0               |    34689925.0               |
| tikv_memtable_size                               | cf, instance, type, db = write, 127.0.0.1:20180, mem-tables, kv                                                               | 2020-02-06 20:26:51.628000 |   345936192.0               |    34593619.2               |    34591688.0               |    34595448.0               |
| heap_mem_usage                                   | instance, job = 127.0.0.1:2379, pd                                                                                            | 2020-02-06 20:26:51.628000 |   234594768.0               |    23459476.8               |    15726376.0               |    30543536.0               |
| tikv_engine_size                                 | instance, type, db = 127.0.0.1:20180, raft, kv                                                                                | 2020-02-06 20:26:51.628000 |   158033300.0               |    15803330.0               |    15801314.0               |    15805322.0               |
| tikv_memtable_size                               | cf, instance, type, db = raft, 127.0.0.1:20180, mem-tables, kv                                                                | 2020-02-06 20:26:51.628000 |   158004880.0               |    15800488.0               |    15798472.0               |    15802480.0               |
| tikv_allocator_stats                             | instance, type = 127.0.0.1:20180, fragmentation                                                                               | 2020-02-06 20:26:51.628000 |   150449744.0               |    15044974.4               |    14866824.0               |    15312544.0               |
| tikv_allocator_stats                             | instance, type = 127.0.0.1:20180, metadata                                                                                    | 2020-02-06 20:26:51.628000 |   113197920.0               |    11319792.0               |    11319792.0               |    11319792.0               |
| tikv_engine_size                                 | instance, type, db = 127.0.0.1:20180, lock, kv                                                                                | 2020-02-06 20:26:51.628000 |    79471680.0               |     7947168.0               |     7942544.0               |     7951584.0               |
| tikv_memtable_size                               | cf, instance, type, db = lock, 127.0.0.1:20180, mem-tables, kv                                                                | 2020-02-06 20:26:51.628000 |    79471680.0               |     7947168.0               |     7942544.0               |     7951584.0               |
| tikv_engine_size                                 | instance, type, db = 127.0.0.1:20180, default, kv                                                                             | 2020-02-06 20:26:51.628000 |    70468030.0               |     7046803.0               |     7046803.0               |     7046803.0               |
| tikv_memtable_size                               | cf, instance, type, db = default, 127.0.0.1:20180, mem-tables, kv                                                             | 2020-02-06 20:26:51.628000 |    68931920.0               |     6893192.0               |     6893192.0               |     6893192.0               |
| tikv_allocator_stats                             | instance, type = 127.0.0.1:20180, dirty                                                                                       | 2020-02-06 20:26:51.628000 |    28433568.0               |     2843356.8               |     2020880.0               |     3548688.0               |
| tikv_block_cache_size                            | cf, instance, db = all, 127.0.0.1:20180, kv                                                                                   | 2020-02-06 20:26:51.628000 |     6222680.0               |      622268.0               |      622268.0               |      622268.0               |
......
```

-- compare 2 different time range metrics and order by ratio.
```sql
> select t1.avg_value /t2.avg_value as ratio, t1.metric_name, t1.label,t1.avg_value,t2.avg_value from 
	METRICS_SUMMARY_BY_LABEL as t1 join METRICS_SUMMARY_BY_LABEL as t2 
where
  t1.metric_name = t2.metric_name and t1.label = t2.label and
  t1.time > "2020-02-06 20:30:00" and t1.time < "2020-02-06 20:35:00" and  
  t2.time > "2020-02-06 20:37:00" and t2.time < "2020-02-06 20:42:00" 
  order by ratio desc;
+--------------------+--------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------+-----------------------------+-----------------------------+
| ratio              | metric_name                                      | label                                                                                                                         | avg_value                   | avg_value                   |
+--------------------+--------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------+-----------------------------+-----------------------------+
| 16.0702113766      | tikv_cop_total_rocksdb_perf_statistics           | instance, req, metric = 127.0.0.1:20180, select, block_cache_hit_count                                                        |          36.7687130543      |           2.28800432009     |
|  5.35837757549     | tikv_scheduler_scan_details                      | instance, tag, req, cf = 127.0.0.1:20180, next, get, write                                                                    |          19.4401938439      |           3.628             |
|  5.0               | tidb_gc_lifetime                                 | instance = 127.0.0.1:10081                                                                                                    |         600.0               |         120.0               |
|  5.0               | tidb_gc_config                                   | instance, type = 127.0.0.1:10081, tikv_gc_life_time                                                                           |         600.0               |         120.0               |
|  4.62741058075     | pd_tso_rpc_duration_0.999                        |                                                                                                                               |           0.00628           |           0.0013571304924   |
|  4.62741058075     | pd_handle_requests_duration_0.999                | type = tso                                                                                                                    |           0.00628           |           0.0013571304924   |
|  4.62741058075     | pd_handle_request_duration_0.999                 | instance, type = 127.0.0.1:10081, tso                                                                                         |           0.00628           |           0.0013571304924   |
|  4.46616284271     | tikv_cop_scan_details                            | instance, tag, req, cf = 127.0.0.1:20180, get, select, lock                                                                   |          27.4402184844      |           6.14402552051     |
|  4.31760073068     | tikv_pd_request_avg_duration                     | instance, type = 127.0.0.1:20180, store_heartbeat                                                                             |           0.0042097538      |           0.000975021560025 |
|  4.15884097494     | pd_client_cmd_duration_0.999                     | instance, type = 127.0.0.1:10081, wait                                                                                        |           0.00793493333333  |           0.00190796747967  |
|  4.15884097494     | pd_tso_wait_duration_0.999                       |                                                                                                                               |           0.00793493333333  |           0.00190796747967  |
|  4.0               | tikv_scheduler_latch_wait_duration_1             | instance, type = 127.0.0.1:20180, scan_lock                                                                                   |           0.002             |           0.0005            |
|  3.809202887       | pd_start_tso_wait_duration_0.999                 |                                                                                                                               |           0.007138304       |           0.00187396266667  |
|  3.66516564812     | tidb_kv_request_duration_0.99                    | instance, type, store = 127.0.0.1:10081, BatchGet, 1                                                                          |           0.0303866666667   |           0.00829066666667  |
|  3.6               | pd_start_tso_wait_duration_1                     |                                                                                                                               |           0.009216          |           0.00256           |
|  3.57937723328     | tikv_grpc_messge_duration_0.99                   | instance, type = 127.0.0.1:20180, kv_batch_get                                                                                |           0.028048          |           0.007836          |
|  3.48548548549     | tikv_scheduler_latch_wait_duration_0.999         | instance, type = 127.0.0.1:20180, scan_lock                                                                                   |           0.001741          |           0.0004995         |
|  3.04999251801     | pd_client_cmd_duration_0.999                     | instance, type = 127.0.0.1:10081, tso                                                                                         |           0.00959474666667  |           0.0031458262963   |
|  3.0370149805      | tidb_kv_request_duration_0.99                    | instance, type, store = 127.0.0.1:10081, Get, 1                                                                               |           0.0133710666667   |           0.00440270026738  |
|  3.0               | pd_tso_rpc_duration_1                            |                                                                                                                               |           0.0096            |           0.0032            |
|  3.0               | pd_handle_request_duration_1                     | instance, type = 127.0.0.1:10081, tso                                                                                         |           0.0096            |           0.0032            |
|  3.0               | pd_handle_requests_duration_1                    | type = tso                                                                                                                    |           0.0096            |           0.0032            |
|  2.82352941176     | pd_tso_wait_duration_1                           |                                                                                                                               |           0.0096            |           0.0034            |
|  2.82352941176     | pd_client_cmd_duration_1                         | instance, type = 127.0.0.1:10081, wait                                                                                        |           0.0096            |           0.0034            |
|  2.78382096671     | tikv_scheduler_command_duration_0.99             | instance, type = 127.0.0.1:20180, batch_get                                                                                   |           0.0101173333333   |           0.00363433333333  |
|  2.78260869565     | pd_client_cmd_duration_1                         | instance, type = 127.0.0.1:10081, tso                                                                                         |           0.0128            |           0.0046            |
|  2.72921951382     | tikv_grpc_messge_duration_0.99                   | instance, type = 127.0.0.1:20180, kv_get                                                                                      |           0.0122831622807   |           0.00450061353383  |
|  2.72691863344     | tikv_grpc_messge_duration_0.999                  | instance, type = 127.0.0.1:20180, kv_batch_get                                                                                |           0.0330448         |           0.012118          |
|  2.70819428995     | tidb_meta_operation_duration_0.99                | instance, type, result = 127.0.0.1:10081, get_ddl_job, ok                                                                     |           0.01328           |           0.00490363636364  |
|  2.65982071049     | tidb_kv_request_duration_0.999                   | instance, type, store = 127.0.0.1:10081, BatchGet, 1                                                                          |           0.0361586666667   |           0.0135944         |
|  2.625             | tikv_grpc_messge_duration_1                      | instance, type = 127.0.0.1:20180, kv_batch_get                                                                                |           0.0336            |           0.0128            |
......
```

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -853,6 +853,10 @@ func createSessionFunc(store kv.Storage) pools.Factory {
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		err = variable.SetSessionSystemVar(se.sessionVars, variable.MaxAllowedPacket, types.NewStringDatum("67108864"))
```
Comment (crazycs520) [inline session/session.go]: In PR https://github.com/pingcap/tidb/pull/11137, `concat_ws` will need this variable, so I set this variable here for internal sql.

File: infoschema/tables.go [hunk 2]
```diff
@@ -107,7 +107,9 @@ const (
 	// TableInspectionResult is the string constant of inspection result table
 	TableInspectionResult = "INSPECTION_RESULT"
 	// TableMetricSummary is a summary table that contains all metrics.
-	TableMetricSummary = "METRIC_SUMMARY"
+	TableMetricSummary = "METRICS_SUMMARY"
+	// TableMetricDetail is a detail table that contains all metrics detail info. It is the table metric_summary by label.
+	TableMetricDetail = "METRICS_SUMMARY_BY_LABEL"
```
Comment (lonng) [inline infoschema/tables.go]: Please update the global variable `TableMetricDetail` name and keep consistent with its value.

File: executor/metric_reader.go [hunk 3]
```diff
@@ -243,3 +243,72 @@ func (e *MetricSummaryRetriever) genMetricQuerySQLS(name, startTime, endTime str
 	}
 	return sqls
 }
+
+// MetricDetailRetriever uses to read metric detail data.
+type MetricDetailRetriever struct {
```
Comment (lonng) [inline executor/metric_reader.go]: Please update the type name.

File: infoschema/tables.go [hunk 4]
```diff
@@ -107,7 +107,9 @@ const (
 	// TableInspectionResult is the string constant of inspection result table
 	TableInspectionResult = "INSPECTION_RESULT"
 	// TableMetricSummary is a summary table that contains all metrics.
-	TableMetricSummary = "METRIC_SUMMARY"
+	TableMetricSummary = "METRICS_SUMMARY"
+	// TableMetricDetail is a detail table that contains all metrics detail info. It is the table metric_summary by label.
```
Comment (lonng) [inline infoschema/tables.go]: Outdated comment.

File: executor/metric_reader.go [hunk 5]
```diff
@@ -243,3 +243,72 @@ func (e *MetricSummaryRetriever) genMetricQuerySQLS(name, startTime, endTime str
...
+	totalRows := make([][]types.Datum, 0, len(infoschema.MetricTableMap))
+	quantiles := []float64{1, 0.999, 0.99, 0.90, 0.80}
+	tps := make([]*types.FieldType, 0, len(e.table.Columns))
+	for _, col := range e.table.Columns {
+		tps = append(tps, &col.FieldType)
+	}
+	startTime := e.extractor.StartTime.Format(plannercore.MetricTableTimeFormat)
+	endTime := e.extractor.EndTime.Format(plannercore.MetricTableTimeFormat)
+	for name, def := range infoschema.MetricTableMap {
+		sqls := e.genMetricQuerySQLS(name, startTime, endTime, def.Quantile, quantiles, def)
```
Comment (lonng) [inline executor/metric_reader.go]: The `e.genMetricQuerySQLS` method accepts the `def`, so I think the `def.Quantile` is unnecessary.

File: executor/metric_reader.go [hunk 6]
```diff
@@ -243,3 +243,72 @@ func (e *MetricSummaryRetriever) genMetricQuerySQLS(name, startTime, endTime str
...
+	}
+	sqls := []string{}
+	for _, quantile := range quantiles {
+		sql := fmt.Sprintf(`select "%[1]s_%[5]v", %[2]s as label,min(time),sum(value),avg(value),min(value),max(value) from metric_schema.%[1]s where time > '%[3]s' and time < '%[4]s' and quantile=%[5]v`,
+			name, labelsColumn, startTime, endTime, quantile)
+		if len(def.Labels) > 0 {
+			sql += " group by " + labels
+		}
+		sqls = append(sqls, sql)
+	}
```
Comment (lonng) [inline executor/metric_reader.go]: It's better to simplify the L284-L313:
```go
    labels := ""
    labelsColumn := `""`
    if len(def.Labels) > 0 {
        labels := "`" + strings.Join(def.Labels, "`, `") + "`"
        labelsColumn = fmt.Sprintf("concat_ws(' = ', '%s', concat_ws(', ', `%s`))", strings.Join(def.Labels, ", "), labels)
    }
	
	if quantile == 0 {
        quantiles = []float64{0}
	}
	sqls := make([]string{}, 0, len(quantiles))
	for _, quantile := range quantiles {
        var sql string
        if quantile == 0 {
            sql := fmt.Sprintf(`select "%[1]s", %[2]s as label,min(time),sum(value),avg(value),min(value),max(value) from metric_schema.%[1]s where time > '%[3]s' and time < '%[4]s'`,
			    name, labelsColumn, startTime, endTime)
        } else {
            sql := fmt.Sprintf(`select "%[1]s_%[5]v", %[2]s as label,min(time),sum(value),avg(value),min(value),max(value) from metric_schema.%[1]s where time > '%[3]s' and time < '%[4]s' and quantile=%[5]v`,
			    name, labelsColumn, startTime, endTime, quantile)
        }
		if len(def.Labels) > 0 {
			sql += " group by " + labels
		}
		sqls = append(sqls, sql)
	}
	return sqls
```

File: executor/metric_reader.go (hunk 5 above)
Comment (crazycs520) [reply executor/metric_reader.go]: great, done.

File: executor/metric_reader.go (hunk 6 above)
Comment (crazycs520) [reply executor/metric_reader.go]: great, done.

**Comment by lonng** [comment]
LGTM

**Pull request merged**